pkg/kv/kvserver/replica_init.go

```diff
@@ -243,6 +253,29 @@ func (r *Replica) loadRaftMuLockedReplicaMuLocked(desc *roachpb.RangeDescriptor)
 	return nil
 }
 
+func (r *Replica) loadUninit(ctx context.Context, desc *roachpb.RangeDescriptor) error {
```

Comment:
Ideally I would like this method gone. This currently has 2 things:
1. Setting `ReplicaState` to empty protos (otherwise things panic on bits of it which are nil).
2. Asserting that an attempt to load state will return the same empty protos.

(1) could be moved to `newUnloadedReplica`. As an optimization, it could point to statically allocated empty protos (if I understand correctly and `state.Desc` is "immutable").
(2) is maybe not needed? Or we could reduce it to only check the `ReplicaID` key.

Reply:
I think it would be fine to "just" give each uninited replica its own loaded trivial state. I would stay away from optimizing the allocation away, it's not worth it even it if would work out ok. Too much conceptual complexity for zero gain (we want to be able to say that two replicas don't share backing state memory even in this trivial case).

It does seem kind of nice to not actually call `.Load()` to load random keys we know shouldn't be there. Maybe we can "just" hard-code:

```go

// stateloader.go
func UninitializedReplicaState(rangeID roachpb.RangeID) kvserver.ReplicaState {
  return ReplicaState{
    // manually filled like in your code below
  }
}

// stateloader_test.go

func TestUninitializedReplicaState(t *testing.T) {
   eng := storage.NewTestInMem()
   defer eng.Close()
   exp, err := stateloader.New(...).Load(&roachpb.RangeDescriptor{RangeID:123})
   require.NoError(t, err)
   act := UninitializedReplicaState(123)
   require.Equal(t, exp, act)
}
```

but actually loading can also discover keys, so maybe we should continue actually loading but asserting that the result matches `UninitializedReplicaState(rangeID)`?

Reply:
Done. Left the assert in place for now.